push_notification: Skip clearing notif for users without push devices.

prakhar1144 opened the pull request:
First commit: https://github.com/zulip/zulip/pull/36121#issuecomment-3444264235

This PR adds check to skip enqueuing events with type="remove" if no push device is registered for the user.

21dedfe40a02cc64a52c1568c22c72eadbcd5cb2 made improvements to skip enqueuing events to send push notif if no push device is registered - it helped to lower the number of events enqueued to *clear* notifs as well because we have overall less number of usermessages with `active_mobile_push_notification` flag set.

But there's still a couple of cases where events to clear can be enqueued when no push device is registered for a user:
* After 21dedfe40a02cc64a52c1568c22c72eadbcd5cb2:
  * event to send push notif is enqueued when push device was present
  * push device unregistered
  * event to clear (type="remove") can be enqueued as usermessages
    with active_mobile_push_notification flag set is present.
* event to send push notif enqueued prior to 21dedfe40a02cc64a52c1568c22c72eadbcd5cb2 results in usermessages with 'active_mobile_push_notification' flag set - so events to clear them can be enqueued post upgrade.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Failing test is [#development help > Issues with CI-tests failing for PR #36096 @ 💬](https://chat.zulip.org/#narrow/channel/49-development-help/topic/Issues.20with.20CI-tests.20failing.20for.20PR.20.2336096/near/2265634)

timabbott commented on zerver/actions/message_flags.py:
```diff
@@ -229,18 +229,28 @@ def do_clear_mobile_push_notifications_for_ids(
         UserMessage.objects.filter(
             message_id__in=message_ids,
-            user_profile_id__in=user_profile_ids,
+            user_profile_id__in=push_device_registered_user_ids,
```
I don't think this is the right layer to do this filtering at. We do want to clear `where_active_push_notification` in our database when the user has no registered devices, whereas this will leak that flag on those users forever?

prakhar1144 commented:
Addressed https://github.com/zulip/zulip/pull/36121#discussion_r2383665602

Also, rebased the PR on top of #26815 as it makes sense to consider that change as well while reviewing this - to see the eventual state.

prakhar1144 commented on zerver/actions/message_flags.py:
```diff
@@ -236,34 +236,54 @@ def do_clear_mobile_push_notifications_for_ids(
-    for user_id, message_id in notifications_to_update:
-        messages_by_user[user_id].append(message_id)
+    for um in usermessage_queryset:
+        messages_by_user[um.user_profile_id].append(um.message_id)
```
> I think you really just need the user_id/message_id pairs here.

Just relying on these two won't help in updating the `flags`. We'll have to write a query in the form of `UserMessage.objects.filter(user_profile_id__in=..., message_id_in=...).update(flags=...)` - which would cost quadratic work.

Reworked it by using the `id` field instead of fetching the `usermessage` objects. Have a look.

timabbott commented on zerver/actions/message_flags.py:
```diff
@@ -256,14 +256,47 @@ def do_clear_mobile_push_notifications_for_ids(
+
+    if usermessages_to_update_ids:
+        BATCH_SIZE = 1000
+        for i in range(0, len(usermessages_to_update_ids), BATCH_SIZE):
```
Seems OK.
